design/baremetal-operator/bulk-set-bios-config.md

```diff
@@ -0,0 +1,389 @@
+
+### Implementation Details/Notes/Constraints
+
+One of the primary drivers of his design was that each vendor uses different
+names and values for similar BIOS Configuration settings. This design does not
+rely on vendor specific checks but instead uses common mechanisms to ensure
+the BIOS Settings are valid:
+
+- copying the names and values from `settingsData` to
+  `vendorSettings` ensures initial values are correct
+- when updating the values in `vendorSettings`, the user can use the
+  limitations in the `settingsData`, e.g. minimum and maximum values
+- before writing to the Ironic clean-steps API, the BMO runs validation checks
+  against the `settingsData` to ensure updated values are valid
+
+In order to interoperate with the individual BIOS Configuration set
+approach the BMO must ensure that the same settings don't get added twice
+to the Ironic clean-steps API. In addition, as each clean step requires a
+reboot the number of clean steps that are added must be managed.
+
+An example of the proposed CRD definition for the BIOS Settings is as follows:
+
+```yaml
+---
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  annotations:
+    controller-gen.kubebuilder.io/version: v0.4.1
+  creationTimestamp: null
+  name: hostfirmwaresettings.metal3.io
+spec:
+  group: metal3.io
+  names:
+    kind: HostFirmwareSettings
+    listKind: HostFirmwareSettingsList
+    plural: hostfirmwaresettings
+    singular: hostfirmwaresettings
+  scope: Namespaced
+  versions:
+  - name: v1alpha1
+    schema:
+      openAPIV3Schema:
+        description: HostFirmwareSettings is Schema for hostfirmwaresettings
+        properties:
+          apiVersion:
+            description: 'APIVersion defines the versioned schema'
+            type: string
+          kind:
+            description: 'Kind is a string value representing the resource'
+            type: string
+          metadata:
+            type: object
+          spec:
+            description: HostFirmwareSettingsSpec defines the desired state
+            properties:
+              vendorSettings:
+                items:
+                  description: describes a setting whose value can be changed
+                  properties:
+                    name:
+                      description: Identifier for the Firmware setting.
+                      type: string
+                    value:
+                      description: Value of the Firmware setting.
+                      type: string
+                  required:
+                  - name
+                  - value
+                  type: object
+                type: array
+            type: object
+          status:
+            description: defines the observed state of HostFirmwareSettings
+            properties:
+              settingsData:
+                items:
+                  description: describes fields read from Ironic
+                  properties:
+                    schema:
```

Comment:
On a second thought, and somehow continuing the discussion on a point previously analysed, I'm starting to think if the schema should really go into a separate, dedicated CRD (with the granularity of at least one per vendor/model). I could envision a couple of benefits with such design:

* Smaller `HostFirmwareSettings` footprint. In scenario with thousands of BMH, currently the same schema will be duplicated thousands time, with a waste of space for a lot of data that will be identical.
* Future support for bulk copy/value drifting. Having a single schema resource for vendor/model (with potentially a golden set of value) could make it easier the work for the future controller responsible to apply the bulk copy

Reply:
This seems like a reasonable change to reduce the memory consumption as yes, the schema is largely fixed per vendor model. We could create a separate CRD referenced from HostFirmwareSettings to contain the schema. When retrieving the data from Ironic the vendor info in BMH can be used to create a new entry in the schema CRD if there is not a reference in HFS.

I'd like to get Zane's thoughts on this. The most recent update I made to the spec did not include any mention of a new CRD but I can add it on next update if we get agreement.

Reply:
How would you do the de-duplication of the schema?

Reply:
I could see it working like this, please correct any misconceptions I have about CRD access:
- We'd add a reference to `settingsData` (in the HFS CRD) to point to a entry in the new schema CRD. 
- The first time the BMO gets BIOS data from Ironic it would use the `manufacturer` and `model` to either access an existing schema entry (and update the reference) or create a new entry and copy the BIOS registry data in API to it, instead of storing this in the HFS CRD as currently proposed (and also update the reference)
- Any future access to Ironic API will not need to update the schema once the schema reference is set, in fact we can even reduce the API bandwidth to Ironic by not use the `?detail=True` flag (which causes Ironic to include BIOS registry info)

Reply:
I don't think we can guarantee that the schemas for a given manufacture+model are all the same, given the potential for different firmware versions and peripherals.
A safe way would be for the name of the Schema resource to contain a hash of a canonical form of the schema (sans actual settings values, obviously). In practice I think that would probably yield good deduplication. Maybe we can do something like `<vendor>-<model>-<first 64? chars of hash>` so the user experience won't be too hideous.

The other challenge is reference counting. Maybe we can also label the HFS CRs with the name of the Schema resource, so we can clean them up when there's nothing left referencing it.

Reply:
Thanks Zane, good suggestions. I've updated the doc to add the hash in that format. I haven't addressed the counting issue but will in a follow-on.

Andrea and Zane - I've added the second CRD to the doc, thanks for all the input!